include/rmm/mr/device/hybrid_memory_resource.hpp

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/detail/aligned.hpp>
+#include <rmm/detail/error.hpp>
+
+#include <cuda_runtime_api.h>
+
+#include <list>
+#include <memory>
+#include <unordered_map>
+#include <cstddef>
+#include <utility>
+#include <algorithm>
+#include <cassert>
+
+
+// forward decl
+using cudaStream_t = struct CUstream_st*;
+
+namespace rmm {
+
+namespace mr {
+
+/**
+ * @brief Allocates memory from one of two allocators selected based on the requested size.
+ *
+ * For example, could use a fixed_multisize_memory_resource for allocations below the size
+ * threshold, and a pool_memory_resource for allocations above the size threshold. This can be 
+ * advantageous if there are many small allocations, because fixed_multisize_memory_resource is much 
+ * faster than pool_memory_resource, and it can reduce fragmentation that can impact the
+ * availability of large contiguous blocks of memory.
+ *
+ * @tparam SmallAllocMemoryResource the memory_resource type to use for small allocations.
+ * @tparam LargeAllocMemoryResource the memory_resource type to use for large allocations.
+ */
+ template <typename SmallAllocMemoryResource, typename LargeAllocMemoryResource>
+class hybrid_memory_resource : public device_memory_resource {
+ public:
+
+  // The default size used to select between the two allocators.
+  static constexpr std::size_t default_threshold_size = 1 << 22; // 4 MiB
+
+  /**
+   * @brief Construct a new hybrid_memory_resource_object that selects between the two
+   * specified memory_resources based on the `threshold_size`.
+   * 
+   * @param small_mr The memory_resource to use for small allocations.
+   * @param large_mr The memory_resource to use for large allocations.
+   * @param threshold_size Allocations > this size (in bytes) use large_mr. Allocations <= this size
+   *                       use small_mr. Must be a power of two.
+   */
+  hybrid_memory_resource(
+    SmallAllocMemoryResource* small_mr,
+    LargeAllocMemoryResource* large_mr,
+    std::size_t threshold_size = default_threshold_size)
```

Comment:
Since it needs to be a power of two, consider making `threshold_size` an exponent of 2.

Reply:
For this class it needn't be a power of two, so I plan to remove that requirement and leave it as a number of bytes rather than an exponent.